planner/core/prepare_test.go

```diff
@@ -264,6 +264,38 @@ func (s *testPrepareSuite) TestPrepareOverMaxPreparedStmtCount(c *C) {
 	}
 }
 
+// unit test for issue https://github.com/pingcap/tidb/issues/8518
+func (s *testPrepareSuite) TestPrepareTableAsNameOnGroupBy(c *C) {
+	defer testleak.AfterTest(c)()
+	store, dom, err := newStoreWithBootstrap()
+	c.Assert(err, IsNil)
+	tk := testkit.NewTestKit(c, store)
+	defer func() {
+		dom.Close()
+		store.Close()
+	}()
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t1")
+	tk.MustExec(`create table t1 (
+		id int(11) unsigned not null primary key auto_increment,
+		partner_id varchar(35) not null,
+		t1_status_id int(10) unsigned
+	  );`)
+	tk.MustExec(`insert into t1 values ("1", "partner1", "10"), ("2", "partner2", "10"), ("3", "partner3", "10"), ("4", "partner4", "10");"`)
+	tk.MustExec("drop table if exists t3")
+	tk.MustExec(`create table t3 (
+		id int(11) not null default '0',
+		preceding_id int(11) not null default '0',
+		primary key  (id,preceding_id)
+	  );`)
+	tk.MustExec(`prepare stmt from 'SELECT DISTINCT t1.partner_id
+	FROM t1
+		LEFT JOIN t3 ON t1.id = t3.id
+		LEFT JOIN t1 pp ON pp.id = t3.preceding_id
+	GROUP BY t1.id ;'`)
+	tk.MustQuery("execute stmt").Sort().Check(testkit.Rows("partner1", "partner2", "partner3", "partner4"))
```

Comment:
I test this with branch master, it will not fail.

Reply:
this my test result in branch master 
```
----------------------------------------------------------------------
FAIL: prepare_test.go:268: testPrepareSuite.TestPrepareTableAsNameOnGroupBy

prepare_test.go:296:
    tk.MustQuery("execute stmt").Sort().Check(testkit.Rows("partner1", "partner2", "partner3", "partner4"))
/Users/hao/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:183:
    tk.c.Assert(errors.ErrorStack(err), check.Equals, "", comment)
... obtained string = "" +
...     "[schema:1146]Table 'test.pp' doesn't exist\n" +
...     "github.com/pingcap/errors.AddStack\n" +
...     "\t/Users/hao/go/pkg/mod/github.com/pingcap/errors@v0.11.0/errors.go:174\n" +
...     "github.com/pingcap/parser/terror.(*Error).GenWithStackByArgs\n" +
...     "\t/Users/hao/go/pkg/mod/github.com/pingcap/parser@v0.0.0-20190121074657-4b899f19591e/terror/terror.go:233\n" +
...     "github.com/pingcap/tidb/infoschema.(*infoSchema).TableByName\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/infoschema/infoschema.go:179\n" +
...     "github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildDataSource\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:1987\n" +
...     "github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildResultSetNode\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:144\n" +
...     "github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildJoin\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:301\n" +
...     "github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildResultSetNode\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:136\n" +
...     "github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildSelect\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:1802\n" +
...     "github.com/pingcap/tidb/planner/core.(*PlanBuilder).Build\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/planner/core/planbuilder.go:188\n" +
...     "github.com/pingcap/tidb/planner.Optimize\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/planner/optimize.go:38\n" +
...     "github.com/pingcap/tidb/planner/core.(*Execute).getPhysicalPlan\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/planner/core/common_plans.go:212\n" +
...     "github.com/pingcap/tidb/planner/core.(*Execute).OptimizePreparedPlan\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/planner/core/common_plans.go:187\n" +
...     "github.com/pingcap/tidb/planner.Optimize\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/planner/optimize.go:54\n" +
...     "github.com/pingcap/tidb/executor.(*Compiler).Compile\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/executor/compiler.go:49\n" +
...     "github.com/pingcap/tidb/session.(*session).execute\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/session/session.go:950\n" +
...     "github.com/pingcap/tidb/session.(*session).Execute\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/session/session.go:912\n" +
...     "github.com/pingcap/tidb/util/testkit.(*TestKit).Exec\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:137\n" +
...     "github.com/pingcap/tidb/util/testkit.(*TestKit).MustQuery\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:182\n" +
...     "github.com/pingcap/tidb/planner/core_test.(*testPrepareSuite).TestPrepareTableAsNameOnGroupBy\n" +
...     "\t/Users/hao/go/src/github.com/pingcap/tidb/planner/core/prepare_test.go:296\n" +
...     "runtime.call32\n" +
...     "\t/usr/local/Cellar/go/1.11.2/libexec/src/runtime/asm_amd64.s:522\n" +
...     "reflect.Value.call\n" +
...     "\t/usr/local/Cellar/go/1.11.2/libexec/src/reflect/value.go:447\n" +
...     "reflect.Value.Call\n" +
...     "\t/usr/local/Cellar/go/1.11.2/libexec/src/reflect/value.go:308\n" +
...     "github.com/pingcap/check.(*suiteRunner).forkTest.func1\n" +
...     "\t/Users/hao/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/check.go:836\n" +
...     "github.com/pingcap/check.(*suiteRunner).forkCall.func1\n" +
...     "\t/Users/hao/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/check.go:730\n" +
...     "runtime.goexit\n" +
...     "\t/usr/local/Cellar/go/1.11.2/libexec/src/runtime/asm_amd64.s:1333"
... expected string = ""
... sql:execute stmt, args:[]
```

Reply:
The test need to specifically enable prepare plan cache, because it is disabled by default. You can reference function `TestPrepareCache` for details.

Reply:
@hhxcc FYI